# Problem Set #5

Ariel Boyarsky

aboyarsky@uchicago.edu

a. Use a multinomial logistic regression model of the following form with the
following linear predictor $η_j$ for j = 1, 2 (the baseline class is j = 3).

In [17]:
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.cross_validation import train_test_split
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix

In [18]:
dta = pd.read_csv("strongdrink.txt")

In [19]:
dta.head()

cultivar   alco  malic   ash   alk  magn  tot_phen  flav  nonfl_phen  \
0         1  14.23   1.71  2.43  15.6   127      2.80  3.06        0.28   
1         1  13.20   1.78  2.14  11.2   100      2.65  2.76        0.26   
2         1  13.16   2.36  2.67  18.6   101      2.80  3.24        0.30   
3         1  14.37   1.95  2.50  16.8   113      3.85  3.49        0.24   
4         1  13.24   2.59  2.87  21.0   118      2.80  2.69        0.39   

   proanth  color_int   hue  OD280rat  proline  
0     2.29       5.64  1.04      3.92     1065  
1     1.28       4.38  1.05      3.40     1050  
2     2.81       5.68  1.03      3.17     1185  
3     2.18       7.80  0.86      3.45     1480  
4     1.82       4.32  1.04      2.93      735

In [20]:
X = dta[['alco', 'malic', 'tot_phen', 'color_int']]
y = dta['cultivar']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state=20)

LR = LogisticRegression()
LR.fit(X_train, y_train)
y_pred = LR.predict(X_test)

print(classification_report(y_test, y_pred))

             precision    recall  f1-score   support

          1       0.92      0.85      0.88        13
          2       0.95      0.95      0.95        21
          3       0.91      1.00      0.95        10

avg / total       0.93      0.93      0.93        44



In [21]:
confusion_matrix = confusion_matrix(y_test, y_pred)
confusion_matrix

array([[11,  1,  1],
       [ 1, 20,  0],
       [ 0,  0, 10]])

In [22]:
LR.coef_

array([[-0.14627245, -0.48477655,  1.51308262,  0.13903482],
       [ 0.3565457 , -0.34661319,  0.64339283, -1.61990007],
       [-0.23172523,  0.95844737, -2.68156609,  1.00492579]])

In [23]:
LR.intercept_

array([-1.94941251,  1.10964888,  0.03887469])

In [28]:
print(len(dta.where(dta.cultivar == 1)))
print(len(dta.where(dta.cultivar == 2)))
print(len(dta.where(dta.cultivar == 3)))

176
176
176


Thus, the coefficents for $j=1$ are -0.146, -0.484, 1.513, 0.139. The coefficents for $j=2$ are 0.356, -0.346, 0.6433, -1.619. 

The error rate for $j=1$ is 0.08, for $j=2$ is 0.05, for $j=3$ is 0.09. It is best at predicting $j=2$. Nope, there is an equal number of observations for each.